File: src/sys/component_manager/tests/events/components/scoped/echo_client.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    anyhow::format_err, fidl_fidl_examples_routing_echo as fecho, fuchsia_async as fasync,
    fuchsia_component::client::connect_to_service, fuchsia_syslog as syslog,
};

#[fasync::run_singlethreaded]
async fn main() {
    syslog::init_with_tags(&["scoped_echo_client"]).expect("failed to init log");
    let echo = connect_to_service::<fecho::EchoMarker>().expect("error connecting to echo");

    let out = echo.echo_string(Some("Hippos rule!")).await.expect("echo_string failed");
    let out = out.ok_or(format_err!("empty result")).expect("echo_string got empty result");

    assert_eq!(out, "Hippos rule!");
}

File: sdk/lib/component/tests/echo_client.cc
// Copyright 2023 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <fidl/fidl.service.test/cpp/fidl.h>
#include <lib/component/incoming/cpp/protocol.h>
#include <lib/syslog/cpp/macros.h>

int main(int argc, const char** argv) {
  FX_LOGS(INFO) << "Starting Echo client";

  auto result = component::Connect<fidl_service_test::Echo>();
  if (result.is_error()) {
    FX_LOGS(ERROR) << "Failed to connect to Echo protocol: " << result.status_string();
    return result.status_value();
  }

  fidl::SyncClient<fidl_service_test::Echo> client(std::move(result.value()));
  auto response = client->EchoString(fidl::Request<fidl_service_test::Echo::EchoString>("Hello!"));
  if (response.is_error()) {
    FX_LOGS(ERROR) << "Didn't receive expected response: "
                   << response.error_value().status_string();
    return response.error_value().status();
  }

  return 0;
}
